src/t_hash.c

```diff
@@ -675,49 +704,34 @@ void hashTypeCurrentFromListpack(hashTypeIterator *hi,
 /* Get the field or value at iterator cursor, for an iterator on a hash value
  * encoded as a hash table. Prototype is similar to
  * `hashTypeGetFromHashTable`. */
-sds hashTypeCurrentFromHashTable(hashTypeIterator *hi, int what) {
+char *hashTypeCurrentFromHashTable(hashTypeIterator *hi, int what, size_t *len) {
     serverAssert(hi->encoding == OBJ_ENCODING_HASHTABLE);
 
     if (what & OBJ_HASH_FIELD) {
-        return entryGetField(hi->next);
-    } else {
-        return entryGetValue(hi->next);
-    }
-}
-
-/* Higher level function of hashTypeCurrent*() that returns the hash value
- * at current iterator position.
- *
- * The returned element is returned by reference in either *vstr and *vlen if
- * it's returned in string form, or stored in *vll if it's returned as
- * a number.
- *
- * If *vll is populated *vstr is set to NULL, so the caller
- * can always check the function return by checking the return value
- * type checking if vstr == NULL. */
-static void hashTypeCurrentObject(hashTypeIterator *hi, int what, unsigned char **vstr, unsigned int *vlen, long long *vll) {
-    if (hi->encoding == OBJ_ENCODING_LISTPACK) {
-        *vstr = NULL;
-        hashTypeCurrentFromListpack(hi, what, vstr, vlen, vll);
-    } else if (hi->encoding == OBJ_ENCODING_HASHTABLE) {
-        sds ele = hashTypeCurrentFromHashTable(hi, what);
-        *vstr = (unsigned char *)ele;
-        *vlen = sdslen(ele);
-    } else {
-        serverPanic("Unknown hash encoding");
+        sds key = entryGetField(hi->next);
+        if (key) *len = sdslen(key);
```

Comment:
Bug.  `if (len) ...`

(Also, I preferred the readability of the original if/else structure)

Reply:
This is not correct. `len` is never NULL while key could be.

Reply:
Ahh, I see.  That was not a pattern I was expecting.  Often C functions support returning optional results.  If the reference pointer is null, the value is not returned.

There is no indication in the comments that key can actually be null.  Is that a real case?  Searching the other code, I see that it often immediately calls sdslen (or similar) on the returned value.  It seems that the result is generally assumed not to be null.  Is null key something new that's being introduced?

If we don't actually expect null here, then this is unnecessary defensive programming which would potentially obfuscate an error (postponing the crash until later).


Re: my previous comment...

In the old code, length was not returned.  Now a length has been added, it looked like it was optional, and if NULL was provided, the length would not be returned.

Would it be better to do this?
```suggestion
        *len = (key) ? sdslen(key) : 0;
```
Or do we want `len` to remain unassigned in this case?

Reply:
I kinda agree with @JimB123 here. There is no way the current code can return NULL so I see no real need to add the check